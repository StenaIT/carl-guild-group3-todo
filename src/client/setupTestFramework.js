
//https://gist.github.com/jsdf/6fc35890e4ed4a219072
var util = require('util');

// keep a reference to the original console methods
var consoleWarn = console.warn;
var consoleError = console.error;

function logToError() {
  throw new Error(util.format.apply(this, arguments).replace(/^Error: (?:Warning: )?/, ''));
}

beforeEach(function() {
  // make calls to console.warn and console.error throw an error
  console.warn = logToError;
  console.error = logToError;
});

afterEach(function() {
  // return console.warn and console.error to default behaviour
  console.warn = consoleWarn;
  console.error = consoleError;
});
