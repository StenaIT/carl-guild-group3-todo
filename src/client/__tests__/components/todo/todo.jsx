//Can use this when React has released v15.4
//https://github.com/facebook/jest/issues/1353
//import React from 'react';
//import renderer from 'react-test-renderer';
//import {shallow} from 'enzyme';

import Todo from '../../../components/todo/todo.jsx';

// We need to include resetmodules until React has released v15.4
beforeEach(() => jest.resetModules());

describe('Todo', () => {
  describe('Snapshot', () => {
    it('should render a view', () => {
      // We need to include require in each test until React has released v15.4
      const React = require('react');
      const renderer = require('react-test-renderer');

      const view = {
          todo: {
            id: '2974203a-8c9f-4e3b-963d-801bbba5e9fb',
            text: 'View this todo',
            completed: false
          },
          editTodo: jest.fn(),
          deleteTodo: jest.fn(),
          completeTodo: jest.fn()
      }

      const component = renderer.create(<Todo { ...view}/>);
      const json = component.toJSON();
      expect(json).toMatchSnapshot();
    });
  });

  describe('Component', () => {
    // We use our consle.warn to throw instead of output it to console.
    require('../../../setupTestFramework.js')

    it('should warn when missing todo data', () => {
      const React = require( 'react');
      const shallow = require('enzyme').shallow;

      expect(() => {shallow(<Todo />) }).toThrowError(/Warning..Failed.prop.type..Required.prop/);
    });

    it('should warn when invalid todo data', () => {
      const React = require( 'react');
      const shallow = require('enzyme').shallow;

      const invalid = {
        todo: {
          id: '26b7dbf3-d3cf-4843-8d42-93e2d3f97886',
          test: "invalid",
        },
        editTodo: jest.fn(),
        deleteTodo: jest.fn(),
        completeTodo: jest.fn()
      }

      expect(() => {shallow(<Todo {...invalid}/>) }).toThrowError(/Warning..Failed.prop.type..Required.prop/);
    });

    it('should be able to change to be editable', () => {
      // We need to include require in each test until React has released v15.4
      const React = require( 'react');
      const shallow = require('enzyme').shallow;

      const edit = {
          todo: {
            id: '1717f1ba-e0d2-468b-996f-d3af4794bf6c',
            text: 'Edit this todo',
            completed: false
          },
          editTodo: jest.fn(),
          deleteTodo: jest.fn(),
          completeTodo: jest.fn()
      }

      const todo = shallow(<Todo {...edit} />);
      expect(todo.find('label').text()).toEqual('Edit this todo');

      todo.find('label').simulate('doubleclick');
      expect(todo.state('editing')).toEqual(true);
    });
  });
});
