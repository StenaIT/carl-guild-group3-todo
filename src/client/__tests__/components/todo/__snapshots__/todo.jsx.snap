exports[`Todo Snapshot should render a view 1`] = `
<div
  className="view">
  <input
    checked={false}
    className="toggle"
    onChange={[Function]}
    type="checkbox" />
  <label
    onDoubleClick={[Function]}>
    View this todo
  </label>
  <button
    className="trash"
    onClick={[Function]} />
</div>
`;
